src/Persistence/Search/FacetSearchIndexFieldsBuilder.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace ProfessionalWiki\WikibaseFacetedSearch\Persistence\Search;
+
+use Exception;
+use ProfessionalWiki\WikibaseFacetedSearch\Application\Config;
+use SearchEngine;
+use SearchIndexField;
+use Wikibase\DataModel\Services\Lookup\PropertyDataTypeLookup;
+
+class FacetSearchIndexFieldsBuilder {
+
+	public function __construct(
+		private readonly SearchEngine $engine,
+		private readonly Config $config,
+		private readonly PropertyDataTypeLookup $dataTypeLookup
+	) {
+	}
+
+	/**
+	 * @return array<string, SearchIndexField>
+	 */
+	public function createFields(): array {
+		$fields = [];
+
+		foreach ( $this->config->getFacets()->asArray() as $facetConfig ) {
+			$name = 'wbfs_' . $facetConfig->propertyId->getSerialization();
+
+			try {
+				$dataTypeId = $this->dataTypeLookup->getDataTypeIdForProperty( $facetConfig->propertyId );
+			} catch ( Exception $e ) {
+				continue;
+			}
+
+			$fieldType = $this->getFieldTypeForDataTypeId( $dataTypeId );
+
+			if ( $fieldType === null ) {
+				continue;
+			}
+
+			$fields[$name] = $this->engine->makeSearchFieldMapping( $name, $fieldType );
+		}
+
+		return $fields;
+	}
+
+	private function getFieldTypeForDataTypeId( string $dataTypeId ): ?string {
+		return match ( $dataTypeId ) {
+			'quantity' => SearchIndexField::INDEX_TYPE_NUMBER,
+			'string' => SearchIndexField::INDEX_TYPE_KEYWORD,
+			'time' => SearchIndexField::INDEX_TYPE_DATETIME,
+			'wikibase-item' => SearchIndexField::INDEX_TYPE_KEYWORD,
+			default => null
```

Comment:
I did some double checking about this approach, in particular if it's better to use Wikibase's internal type, and it seems good as it is.

Here is what I did in Semantic Wikibase, which has a fuller list of types: https://github.com/ProfessionalWiki/SemanticWikibase/blob/master/src/Translation/PropertyTypeTranslator.php#L31

Reply:
Which makes me think we could go further than extracting a getFieldTypeForPropertyId method as suggested above, and instead make that a service.